auth/auth-router.js

```diff
@@ -0,0 +1,70 @@
+const express = require('express');
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+const usersModel = require('../users/users-model');
+const secrets = require('../config/secrets');
+
+const router = express.Router();
+
+router.post('/register', async (req, res, next) => {
+  try {
+    const user = await usersModel.add(req.body);
+    res.status(201).json(user);
+  } catch (error) {
+    next(error);
+  }
+});
+
+router.post('/login', validateUser, async (req, res, next) => {
+  try {
+    const { username, password } = req.body;
+    const user = await usersModel.getBy({ username }).first();
+    const passwordValid = await bcrypt.compare(password, user.password);
+
+    if (user && passwordValid) {
+      const token = signToken(user);
+
+      res.status(200).json({
+        message: 'Logged in',
+        user: user.id,
+        token,
+      });
```

Comment:
I believe this is the pattern that Jason did in class, but I believe there is 1 logical error in it.  If you try to login a user with a username that doesn't exist.  It will successfully return null.  The issue will then be on line 22 with something like `can't get password of undefined` and it will proceed to catch the error instead of saying `invalid credentials...`

To mitigate the problem, you need to have an if/else block and test if user is null first.  Then if user isn't null, you can check password.

From a security standpoint, you would also make the decision when searching the user by username.  If username doesn't exist, do you say `invalid credentials` or do you say `user doesn't exist?`.  I personally would use `user doesn't exist` as it's not as important for this app to be highly secure.  This makes it easier to understand what went wrong for the user.